internal/filter/operators.go

```diff
@@ -0,0 +1,295 @@
+		default:
+			return nil, fmt.Errorf("unknown operator in expression: %s", e)
+		}
+	default:
+		panic(fmt.Sprintf("unknown attribute type: %s", typ))
+	}
+}
+
+func cmpDataTime(ref time.Time, cmp func(v, ref time.Time) bool) func(interface{}) bool {
+	return func(i interface{}) bool {
+		date, ok := i.(string)
+		if !ok {
+			return false
+		}
+		value, err := datetime.Parse(date)
+		if err != nil {
+			return false
+		}
+		return cmp(value, ref)
+	}
+}
+
+func cmpInteger(ref int, cmp func(v, ref int) bool) func(interface{}) bool {
+	return func(i interface{}) bool {
+		value, ok := toInt(i)
+		if !ok {
+			return false
+		}
+		return cmp(value, ref)
+	}
+}
+
+func cmpDecimal(ref float64, cmp func(v, ref float64) bool) func(interface{}) bool {
+	return func(i interface{}) bool {
+		value, ok := toFloat(i)
+		if !ok {
+			return false
+		}
+		return cmp(value, ref)
+	}
+}
+
+func cmpBool(ref bool, cmp func(v, ref bool) bool) func(interface{}) bool {
+	return func(i interface{}) bool {
+		value, ok := i.(bool)
+		if !ok {
+			return false
+		}
+		return cmp(value, ref)
+	}
+}
+
+// The entire operator value must be a substring of the attribute value for a match.
+func strComp(ref interface{}, attr schema.CoreAttribute, cmp func(v, ref string) bool) (func(interface{}) bool, error) {
+	switch typ := attr.AttributeType(); typ {
+	case "boolean":
+		ref, ok := ref.(bool)
+		if !ok {
+			return nil, fmt.Errorf("a boolean attribute needs to be compared to a boolean")
+		}
+		return func(i interface{}) bool {
+			value, ok := i.(bool)
+			if !ok {
+				return false
+			}
+			return cmp(fmt.Sprintf("%t", value), fmt.Sprintf("%t", ref))
+		}, nil
+	case "decimal":
+		ref, ok := toFloat(ref)
+		if !ok {
+			return nil, fmt.Errorf("a decimal attribute needs to be compared to a float/int")
+		}
+		return func(i interface{}) bool {
+			value, ok := toFloat(i)
+			if !ok {
+				return false
+			}
+			// fmt.Sprintf("%f") would give them both the same precision.
+			return cmp(fmt.Sprint(value), fmt.Sprint(ref))
+		}, nil
+	case "integer":
+		ref, ok := toInt(ref)
+		if !ok {
+			return nil, fmt.Errorf("a integer attribute needs to be compared to a int")
+		}
+		return func(i interface{}) bool {
+			value, ok := toInt(i)
+			if !ok {
+				return false
+			}
+			return cmp(fmt.Sprintf("%d", value), fmt.Sprintf("%d", ref))
+		}, nil
+	case "binary", "reference", "dateTime", "string":
+		ref, ok := ref.(string)
+		if !ok {
+			return nil, fmt.Errorf("a %s attribute needs to be compared to a string", typ)
+		}
+		if attr.CaseExact() {
+			return func(i interface{}) bool {
+				value, ok := i.(string)
+				if !ok {
+					return false
+				}
+				return cmp(value, ref)
+			}, nil
+		}
+		return func(i interface{}) bool {
+			value, ok := i.(string)
+			if !ok {
+				return false
+			}
+			return cmp(strings.ToLower(value), strings.ToLower(ref))
+		}, nil
+	default:
+		panic(fmt.Sprintf("unknown attribute type: %s", typ))
+	}
+}
```

Comment:
Can the type conversions (e.g. `toInt`) in these functions ever fail?

Reply:
Indeed, these would never fail, since this is checked in advance by the `createCompareFunction` func.

Reply:
Yes, these can fail since we have no control what will be passed as `i`. 
This is just the value extracted from the resource `map[string]interface{}`.

Reply:
> Yes, these can fail since we have no control what will be passed as `i`.
> This is just the value extracted from the resource `map[string]interface{}`.

I see. Is it okay to just return `false` in the comparison functions if those conversions fail?